### PR TITLE
User Libraries

### PR DESCRIPTION
**Pull request by trauman413**
## Jira Ticket Number
https://gaming-backlog.atlassian.net/browse/GB-64

## Short description
Adds support for libraries for users. 

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Documentation
- [ ] Other (please specify)

### CODE REVIEW

File: src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/controllers/LibraryController.kt [hunk 1]
```diff
@@ -14,6 +14,7 @@ import java.util.*
 @RequestMapping("/libraries")
 class LibraryController(private val libraryService: LibraryService) {
 
+  // TODO: the following two functions can be removed given they exist now in UserController
```
Comment (ksadoff) [inline src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/controllers/LibraryController.kt]: Can we just remove them now?

File: src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/controllers/UserController.kt [hunk 2]
```diff
@@ -54,4 +58,35 @@ class UserController(private val userService: UserService) {
...
+
+  @CrossOrigin(origins = ["http://localhost:3000"])
+  @PatchMapping("/{userId}/libraries/{libraryId}")
+  fun deleteUserLibrary(
+    @PathVariable("userId") userId: String,
+    @PathVariable("libraryId") libraryId: String
+  ): ResponseEntity<UserResponse> {
+    val userRespsonse = userService.deleteUserLibrary(userId, libraryId)
+      ?: return ResponseEntity(HttpStatus.NOT_FOUND)
+    return ResponseEntity.ok(userRespsonse)
```
Comment (ksadoff) [inline src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/controllers/UserController.kt:90]: What if the library doesn't exist? Will the service method throw an error or do we swallow those?

File: src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/services/LibraryService.kt [hunk 3]
```diff
@@ -82,12 +82,12 @@ class LibraryService (
   }
 
   fun convertLibraryToResponse(library: Library): LibraryResponse {
-    if (library.games.size > 0) {
+    return if (library.games.size > 0) {
```
Comment (ksadoff) [inline src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/services/LibraryService.kt:85]: Oooh slick!

File: src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/services/UserService.kt [hunk 4]
```diff
@@ -32,7 +34,8 @@ class UserService(
     val user = User(
       displayName = userRequest.displayName,
       password = userRequest.password,
-      email = userRequest.email
+      email = userRequest.email,
+      libraries = ArrayList() // TODO: decide if we want to have a few generic libraries created upon account creation
```
Comment (ksadoff) [inline src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/services/UserService.kt]: I think we agreed on a default one of just all the games you add? I think it makes sense to just initialize a default library for everyone, avoids some null pointers

File: src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/services/UserService.kt [hunk 5]
```diff
@@ -58,6 +61,32 @@ class UserService(
   }
 
   fun convertUserToResponse(user: User): UserResponse {
-    return UserResponse(user.id, user.displayName, user.email)
+    // might want to have better error handling later without !!
```
Comment (ksadoff) [inline src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/services/UserService.kt:65]: What's the goal of this right now? Just to make sure it's not null?

File: src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/controllers/LibraryController.kt (hunk 1 above)
Comment (trauman413) [reply src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/controllers/LibraryController.kt]: yep! i just didn't want to screw up the frontend functionality but it doesn't really matter since we aren't live and are planning on fixing the endpoints anyway

File: src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/controllers/UserController.kt (hunk 2 above)
Comment (trauman413) [reply src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/controllers/UserController.kt:90]: We currently swallow those. In https://gaming-backlog.atlassian.net/browse/GB-54 we can revisit this and maybe come up with better error handling

File: src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/services/UserService.kt (hunk 4 above)
Comment (trauman413) [reply src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/services/UserService.kt]: right! will fix that. 

We may also want to have set empty libraries like Backlog etc. since that's common but that can be discussed down the line

File: src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/services/UserService.kt (hunk 5 above)
Comment (trauman413) [reply src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/services/UserService.kt:65]: Exactly. !! basically just overrides null handling. It's not great practice to use throughout code, but since we're doing a refactor/error handling focus later, I figured I'd defer it

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt [hunk 6]
```diff
@@ -0,0 +1,157 @@
+package com.gamingbacklog.api.gamingbacklogapi.unit.services
```
Comment (trauman413) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt:1]: Note: these tests aren't complete. The remainder of these tests will be added in https://gaming-backlog.atlassian.net/browse/GB-79

File: src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/models/responses/UserResponse.kt [hunk 7]
```diff
@@ -7,12 +7,22 @@ data class UserResponse(
...
+  val email: String,
+  val libraries: List<LibraryResponse>?
 ) : Response() {
   constructor(id: String, displayName: String, email: String) : this(
     id = id,
     displayName = displayName,
     password = null,
-    email = email
+    email = email,
+    libraries = null
```
Comment (ksadoff) [inline src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/models/responses/UserResponse.kt]: So are we going to allow libraries to be null? Shouldn't we always initialize a user with one?

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+      mockGetSingleLibrary("Backlog")
+      mockDeleteLibrary()
+      mockSave(user)
+      val result = userService.deleteUserLibrary(user.id, libraryId)
+      assertNotNull(result)
+      assertEquals(ArrayList<LibraryResponse>(), result?.libraries)
+    }
+
+    @Test
+    fun shouldDeleteNotFoundLibrary() {
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt]: What is this actually testing? When we don't find the library, we're still trying to delete it?

**Review by ksadoff — COMMENTED** [review]
Two last questions

File: src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/models/responses/UserResponse.kt (hunk 7 above)
Comment (trauman413) [reply src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/models/responses/UserResponse.kt]: In `UserService` when we create a User (method `create()`, the library gets initialized there. This constructor isn't currently in use, so I'll remove it.

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt (hunk 8 above)
Comment (trauman413) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt]: We need better error handling throughout tbh. This is basically to show that there is no error when we attempt to delete a library not found. I'm going to remove this and save these edge cases for when we have better error handling (we love pushing things back to this long-awaited error handling ticket 🙃  )

File: src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/controllers/UserController.kt [hunk 9]
```diff
@@ -54,4 +58,35 @@ class UserController(private val userService: UserService) {
...
+    return ResponseEntity.ok(userResponse)
+  }
+
+  @CrossOrigin(origins = ["http://localhost:3000"])
+  @PatchMapping("/{userId}/libraries/{libraryId}")
+  fun deleteUserLibrary(
+    @PathVariable("userId") userId: String,
+    @PathVariable("libraryId") libraryId: String
+  ): ResponseEntity<UserResponse> {
+    val userRespsonse = userService.deleteUserLibrary(userId, libraryId)
```
Comment (ksadoff) [inline src/main/kotlin/com/gamingbacklog/api/gamingbacklogapi/controllers/UserController.kt]: Super nit and probably just worth addressing as a driveby somewhere else but :')
```suggestion
    val userResponse = userService.deleteUserLibrary(userId, libraryId)
```

**Review by ksadoff — APPROVED** [review]
LGTM!
